### PR TITLE
Ensure `GlobalTransform` consistency when `Parent` is spawned without `Children`

### LINKED ISSUE #3329: `GlobalTransform` not being updated
## Bevy version

ffecb05a0ab423b46ab991b49c54516bcaf2ea7b

## Operating system & version

Linux

## What you did

Link to repo [here](https://github.com/yilinwei/bevy-issue).

The only file is `src/main.rs`. The bevy submodule is simply to add some extra logging.

## What you expected to happen

The `TransformPlugin` has two systems `parent_update` and `transform_propagate`. `parent_update` is always meant to run before `transform_propagate` [according to the run criteria](https://github.com/bevyengine/bevy/blob/ffecb05a0ab423b46ab991b49c54516bcaf2ea7b/crates/bevy_transform/src/lib.rs#L37).

If an `Entity` has a `Parent` but no `Children`, `parent_update` [creates one](https://github.com/bevyengine/bevy/blob/ffecb05a0ab423b46ab991b49c54516bcaf2ea7b/crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs#L58). The `transform_propagate` then queries from the [root nodes](https://github.com/bevyengine/bevy/blob/ffecb05a0ab423b46ab991b49c54516bcaf2ea7b/crates/bevy_transform/src/transform_propagate_system.rs#L12) and updates the `GlobalTransform`.

Whether you insert a `Children` component [shouldn't matter](https://github.com/yilinwei/bevy-issue/blob/79e48b8a47e8baf0d7f8aac73df3ee6b66ec5f00/src/main.rs#L15). The `Children` should be inserted by the `parent_update` and the `transform_propagate` should pick up the inserted `Children`.

## What actually happened

Whether you insert the `Children` does matter, if you comment the line out you get the `unexpected matrix:` log line, whereas otherwise you don't.

Adding some more `println!` statements, the root node query doesn't pick up `Children` at all - it's empty, but you can clearly see the new `Children` in the later query and adding the `println!` directly where the `Children` is added shows that it does hit that part of the code.

### PR DESCRIPTION
**Pull request by yilinwei**
# Objective

Closes #3329

# Solution

`parent_update` and `transform_propagate` run in the same stage but `parent_update` can spawn `Children`. This means that the `system` can enter an inconsistent state where the `GlobalTransform` has not been updated on `Children` when spawning an `Entity` where the `Parent` does not have an existing `Children` component.

Introduce a marker trait, `DirtyParent`, so that the system will revert to the correct state the next time the systems run.

### CODE REVIEW

File: crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs
Comment (yilinwei) [inline crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs:125]: Oops, will delete this

**Comment by alice-i-cecile** [comment]
Thanks for the fix :) I'll dig into the details now. Could you swap the title of this PR over to something more descriptive, like "Ensure GlobalTransform consistency when spawning children"?

File: crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs
Comment (alice-i-cecile) [inline crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs:70]:
```suggestion
        // Mark the entity with a `DirtyParent` component so that systems which run
```

File: crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs
Comment (alice-i-cecile) [inline crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs:71]:
```suggestion
        // in the same stage have a way to query for a missed update
```

File: crates/bevy_transform/src/transform_propagate_system.rs
Comment (alice-i-cecile) [inline crates/bevy_transform/src/transform_propagate_system.rs:22]: Like you suggested on Discord, I think this is going to be cleaner with an `Option<&DirtyParent>` query parameter, rather than creating a new query entirely.

**Comment by alice-i-cecile** [comment]
Looks good! Thanks for the fix; the code quality, tests and debugging are excellent for a new contributor. I left a couple of nits; once those are cleaned up this will have my approval.

File: crates/bevy_transform/src/transform_propagate_system.rs
Comment (yilinwei) [reply crates/bevy_transform/src/transform_propagate_system.rs]: I've switched this to use an `Or` and the same `changed_transform_query`, wdyt?

**Review by alice-i-cecile — APPROVED** [review]
Looks good to me :) The `Or` filter is nicer, I'm glad you suggested that idea.

**Comment by Bobox214** [comment]
Not a usual reviewer, but I have reviewed the changes and approves them.

File: crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs
Comment (IceSentry) [inline crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs:71]: I feel like this comment is more appropriate on the struct declaration. Right now the only explanation as to what `DirtyParent` means is here but the struct itself has no documentation.

**Comment by Davier** [comment]
Could it be done by using `set_changed()` on the dirty parent's transform instead of inserting a new component?

File: crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs
Comment (yilinwei) [reply crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs:71]: I'm not certain the struct is a great place for this particular bit of documentation - `DirtyParent` is a quirk of both the `parent_update_system` and how the stages are currently run; the comment is there because it's not obvious why you'd add it in that particular code branch.

As to the wider point, I don't know what the documentation should achieve. The problem is quite a niche one - you need to spawn your components in a particular way and even then, you'll only touch `DirtyParent` in the specific case that you're running a system in the same stage. If you're looking at `DirtyParent` directly, it's likely that you'll want to read through the code itself. If not, then arguably it should be part of the documentation on the whole hierarchy system.

**Comment by yilinwei** [comment]
@Davier I don't think so, but I'm not too familiar with the semantics of `set_changed`.

- The `parent_update_system` runs before the `transform_propagate_system` (which is where we add the new `Children` and mark it with `DirtyParent`).
- The `Transform` is already changed for that particular frame.
- The `Children` component is not visible on that frame.

**Comment by Davier** [comment]
> @Davier I don't think so, but I'm not too familiar with the semantics of `set_changed`.

You're right, I had misunderstood the issue

**Review by mirkoRainer — APPROVED** [review]
LGTM. Test passes as expected and change is fairly straightforward. 👍

File: crates/bevy_transform/src/transform_propagate_system.rs
Comment (DJMcNab) [inline crates/bevy_transform/src/transform_propagate_system.rs:17]: I think `Changed<Children>` would be a better filter here; in that case we could avoid the entire `DirtyParent` bookkeeping. 

I'm not actually sure that (without doing that) any changes to `Children` which didn't also change the child's parent would result in transforms being updated; unless there is a system similar to `update_parents` but the other way around.

**Review by DJMcNab — COMMENTED** [review]
I like the test, but I agree that the actual mechanism of using a temporary component for this is unneccessary.

File: crates/bevy_transform/src/transform_propagate_system.rs
Comment (yilinwei) [reply crates/bevy_transform/src/transform_propagate_system.rs:17]: I did have a discussion about this when implementing it. The options were:

1. Add a marker component
2. Remove the automatic insertion of the `Children` resource, so that it fails in a way which is obvious to the user (how `Transform` and `GlobalTransform` behave)
3. Recalculate when the `Children` change.

My preference was 1. or 2. The reasoning was that `Children` feels like a general `Component` since it's first-class in the API and using it when there are `Parent`/`Child` relationships is encouraged.

IMO, it shouldn't really trigger a `Transform` update; not merely for optimization purposes, but I don't like the idea of the user receiving the `GlobalTransform` update events each time a child is added or removed. I think that would be confusing.

Comment (DJMcNab) [reply crates/bevy_transform/src/transform_propagate_system.rs:17]: I suppose what actually needs to happen is that `Changed<Children>` needs to be reacted to by re-calculating the children, but 'self' doesn't need to be recalculated (unless it already needed to be). The tree has changed, so the transformations of the subtree need updating.

Comment (yilinwei) [reply crates/bevy_transform/src/transform_propagate_system.rs:17]: I don't quite follow; could you please elaborate?

I think in general `Changed<Children>` is too coarse a filter; under normal operation if a child is changed, there shouldn't be a reason to recalculate the `GlobalTransform` of it's siblings. It's only in this specific case, where we "lose" the insertion since we're in the same stage that we need to react to it.

Comment (DJMcNab) [reply crates/bevy_transform/src/transform_propagate_system.rs:17]: Well my point is: if you add a chilld, how does the child's transform get updated? Neither the parent or child's transforms have changed, so that filter won't activate; so changed will be `false`, so it won't activate.

If you don't want to update 'self''s transform then, the way to implement this would be to add a `changed_children` query.
If this solution is implemented, it cleanly sidesteps the need for the messy `DirtyParent` component.

Comment (yilinwei) [reply crates/bevy_transform/src/transform_propagate_system.rs:17]: OK I don't understand the solution at all; at least I can't think of a way of doing it without recalculating the siblings which haven't changed.

This is how I think that the algorithm currently works.

1. Go to the root nodes
2. Walk down each node checking if the `Transform` has changed
3. If the `Transform` has changed, then update the `GlobalTransform` _and_ the `GlobalTransform` of the `Children`

In the problematic case, the `Children` has been created in this stage so it doesn't see the child when walking down the tree. In the frame afterwards, neither the parent nor child have an updated `Transform` but has a `Changed<Children>` which would evaluate to `true`.

The `Changed<Children>` query:

1. Can't distinguish _which_ child has changed, which means we'd have to update _all_ the siblings since we need to mark the parent. Note, this is _also_ true of `DirtyParent`, but since it only lasts for a single frame, that doesn't matter since any changes _during_ that frame needs to be calculated anyway.
2. Will also evaluate to true when mutating `Children`.

I cannot see how a single `Changed<Children>` would be able to sidestep both the issues?

**Comment by yilinwei** [comment]
I _think_ there's another method without the marker component. We could do a dummy update of the `Transform `using `commands.insert` which would apply on the same frame as the children insertion; but that would overwrite any user mutations for that frame.

**Comment by alice-i-cecile** [comment]
I'm going to block this on seeing if we can get @DJMcNab's suggestion to do this without a temporary component to work.

Reviewers: if you can submit a PR to this branch to fix that, it would be greatly appreciated.

**Comment by alice-i-cecile** [comment]
Closing in favor of #4608 <3

**Pull request closed without merging**
